components/brave_new_tab/new_tab_page.mojom

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+module brave_new_tab.mojom;
+
+struct BraveBackground {
+  string author;
+  string image_url;
+  string link;
+};
+
+enum SelectedBackgroundType {
+  kBrave,
+  kCustom,
+  kSolid,
+  kGradient,
+};
+
+struct SelectedBackground {
+  SelectedBackgroundType type;
+  string value;
+};
+
+struct SponsoredImageLogo {
+  string alt;
+  string destination_url;
+  string image_url;
+};
+
+struct SponsoredImageBackground {
+  string wallpaper_id;
+  string creative_instance_id;
+  string campaign_id;
+  string image_url;
+  SponsoredImageLogo? logo;
+};
+
+// WebUI-side handler for notifications from the browser.
+interface NewTabPage {
+
+  // Called when a background-related profile preference has been updated.
+  OnBackgroundPrefsUpdated();
+
+};
+
+// Browser-side handler for requests from the WebUI page.
+interface NewTabPageHandler {
+
+  // Sets the NewTabPage remove interface that will receive notifications from
+  // the browser.
+  SetNewTabPage(pending_remote<NewTabPage> page);
+
+  // Gets or sets whether the user has enabled background images or colors on
+  // the new tab page.
+  GetBackgroundsEnabled() => (bool enabled);
+  SetBackgroundsEnabled(bool enabled) => ();
+
+  // Returns a value indicating whether user-selectable backgrounds are
+  // supported for the current profile.
+  GetBackgroundsCustomizable() => (bool customizable);
+
+  // Gets or sets whether the user has enabled sponsored background images.
+  GetSponsoredImagesEnabled() => (bool enabled);
+  SetSponsoredImagesEnabled(bool enabled) => ();
+
+  // Returns the current collection of Brave background images.
+  GetBraveBackgrounds() => (array<BraveBackground> backgrounds);
+
+  // Returns the list of custom background images supplied by the user.
+  GetCustomBackgrounds() => (array<string> backgrounds);
+
+  // Returns the user-selected or default background.
+  GetSelectedBackground() => (SelectedBackground? background);
+
+  // Returns sponsored image background info, if a sponsored image should be
+  // displayed to the user.
+  GetSponsoredImageBackground() => (SponsoredImageBackground? background);
+
+  // Saves the provided background as the user's selected background.
+  SelectBackground(SelectedBackground background) => ();
+
+  // Displays a file select dialog for selecting custom background images.
+  ShowCustomBackgroundChooser() => (bool images_selected);
+
+  // Adds the custom background images selected by the custom image background
+  // chooser.
+  AddCustomBackgrounds() => ();
```

Comment:
This is a bit non-obvious. The operation is split into two: first, the front end calls `ShowCustomBackgroundChooser` in order to allow the user to choose some files. It returns a boolean value indicating whether images were selected or not. The page handler temporarily stores the list of files selected. The front-end then calls `AddCustomBackgrounds`, which adds in any files previously selected.

I split this into two methods so that the front-end could show a spinner while files are being added (but not while the file selection UI was open). I suppose that `ShowCustomBackgroundChooser` could return a list of paths to the front-end, but it seems a bit sketchy for the front-end to know file paths?

If we leave it the way it is, I'll add better comments.

Reply:
Ah okay - yeah I guess its an unfortunate side effect of showing the dialog through C++ rather than an `input[type=file]` (which I guess we can't do in WebUI?)

So the reason this doesn't take any arguments, or return anything is because its basically telling the backend we're ready to proceed after we turn on the spinner, and then we're notified of the new backgrounds through the same channel we're notified of everything else background related (the background observer).

So I think this would be slightly cleaner without this `AddCustomBackgrounds` call, as we always need to do it if `images_selected` is true, and it makes it a bit harder to reason about.

What if we did something like the following:

```cpp
// C++ side
void ShowCustomBackgroundChooser(Cb callback) {
    auto results = PickFiles();
    std::move(callback).Run(!results.empty());

    CopyImages();
    NotifyBackgroundsChanged()
}
```

```ts
// Typescript side:
const onPickClick = async () => {
    const { picked } = await api.showCustomBackgroundChooser()

    // If we picked something, show a spinner till the backgrounds change
    if (picked) setLoading(true)
}

useEffect(() => {
    setLoading(false)
}, [customBackgrounds])
```

Reply:
This is a good idea - thanks!